File: twopointers/2161-partition-array-according-to-given-pivot.py
from typing import List
import unittest

# https://leetcode.com/problems/partition-array-according-to-given-pivot/

# python3 -m unittest twopointers/2161-partition-array-according-to-given-pivot.py


class Solution(unittest.TestCase):
    # Approach: Additional space
    # Time: O(n)
    # Space: O(n)
    def pivotArray(self, nums: List[int], pivot: int) -> List[int]:
        n = len(nums)
        ans = [pivot] * n
        left, right = 0, n - 1
        for idx in range(n):
            if nums[idx] < pivot:
                ans[left] = nums[idx]
                left += 1
            if nums[n - idx - 1] > pivot:
                ans[right] = nums[n - idx - 1]
                right -= 1
        return ans

    def test(self):
        for nums, pivot, expected in [
            ([9, 12, 5, 10, 14, 3, 10], 10, [9, 5, 3, 10, 10, 12, 14]),
            ([-3, 4, 3, 2], 2, [-3, 2, 4, 3]),
        ]:
            output = self.pivotArray(nums, pivot)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: twopointers/2161-partition-array-according-to-given-pivot.go
package twopointers

// https://leetcode.com/problems/partition-array-according-to-given-pivot/

// Approach: In-place
// Time: O(n)
// Space: O(n)
func pivotArray(nums []int, pivot int) []int {
	n := len(nums)
	left := 0

	for idx := range n {
		if nums[idx] < pivot {
			nums[left] = nums[idx]
			left++
		} else if nums[idx] > pivot {
			nums = append(nums, nums[idx])
		}
	}

	for ; left < n; left++ {
		nums[left] = pivot
	}

	greater := len(nums) - n
	copy(nums[n-greater:], nums[n:])
	nums = nums[:n]

	return nums
}

// // Approach: Additional space
// // Time: O(n)
// // Space: O(n)
// func pivotArray(nums []int, pivot int) []int {
// 	var n = len(nums)
// 	var ans = make([]int, n)
// 	var left, right = 0, n - 1
// 	for idx := 0; idx < n; idx++ {
// 		if nums[idx] < pivot {
// 			ans[left] = nums[idx]
// 			left++
// 		}
// 		if nums[n-idx-1] > pivot {
// 			ans[right] = nums[n-idx-1]
// 			right--
// 		}
// 	}
// 	for ; left <= right; left++ {
// 		ans[left] = pivot
// 	}
// 	return ans
// }
